### ISSUE TITLE
Would it be possible to cache """getters""" to avoid calling them N times during render?

### ISSUE THREAD

**Issue by AllanOricil** [salesforce/lwc#2266]
**Is your feature request related to a problem? Please describe.**
In Vue js when I have a component like the one below, myComputedData is cached to avoid calling it 4 times during the same renderer. In LWC the closest thing we have to a computed prop is an es6 getter, but using a getter I can see it is invoked 4 times.

Vue
````html
<template>
    <div>
        <div v-if="myComputedValue">Show something</div>
        <div v-if="myComputedValue">Show something</div>
        <div v-if="myComputedValue">Show something</div>
        <div v-if="myComputedValue">Show something</div>
    </div>
</template>
<script>
export default {
computed: {
 myComputedValue(){
  return true; //replace this with a computation that would take several milliseconds
 }
}
}
</script>
````

LWC
````html
<template>
    <div>
        <div if:true={myComputedValue}>Show something</div>
        <div if:true={myComputedValue}>Show something</div>
        <div if:true={myComputedValue}>Show something</div>
        <div if:true={myComputedValue}>Show something</div>
    </div>
</template>
````
````js
import { LightningElement, api } from 'lwc';

export default class Test extends LightningElement {

    @api
    set myComputedValue(value){
        console.log(value);    
    }

    get myComputedValue(){
        return true;//replace this with a computation that would take several milliseconds
    }

}
````

**Describe the solution you'd like**
I would like to have a way to cache getters. I thought about adding a new annotation called @cache or make it in a way that we developers don't need to care how it is implemented.

**Describe alternatives you've considered**
No alternatives

**Comment by abdulsattar** [comment]
Hi @AllanOricil, you can try moving the expensive computation to the setter and cache it:

```js
import { LightningElement, api } from 'lwc';

export default class Test extends LightningElement {

    @api
    set myComputedValue(value){
        this._value = true; //replace this with a computation that would take several milliseconds
        console.log(value);
    }

    get myComputedValue(){
        return this._value;
    }

}
```

**Comment by AllanOricil** [comment]
@abdulsattar good point. But then I would not need the getter at all. I could just use `_value` in the template, like this

````html
<template>
    <div>
        <div if:true={_value}>Show something</div>
        <div if:true={_value}>Show something</div>
        <div if:true={_value}>Show something</div>
        <div if:true={_value}>Show something</div>
    </div>
</template>
````

````js
import { LightningElement, api, track } from 'lwc';

export default class Test extends LightningElement {
    
    @api 
    set a(newA){
         this._a = newA; 
         this._value = this._a * this._b;
    }

    get a(){ 
       return this._a
   }  

   @api 
    set b(newB){
         this._b = newB; 
         this._value = this._a * this._b;
    }

    get b(){ 
       return this._b
   } 

    //get/set c, get/set d, etc
    ....
   
@track
   _a;

   @track
   _b;

   //_c, _d, .... etc
    ....

   //depends on the _a and _b
   _value;

}
````

but you can see that this_value = this._a * this._b keeps repeating everywhere, making the developer having to declare getters and setters for each variable _value depends on. If I had some sort of cached computed prop, I would not need to keep repeating things and I could have only one "getter" for value, like. 

Obs: I'm saying getter but it is not a getter, it is something that once computed is cached to avoid repeating code like shown above.

````html
<template>
    <div>
        <div if:true={_value}>Show something</div>
        <div if:true={_value}>Show something</div>
        <div if:true={_value}>Show something</div>
        <div if:true={_value}>Show something</div>
    </div>
</template>
````

````js
import { LightningElement, api, track } from 'lwc';

export default class Test extends LightningElement {
    
    @api 
    a;
    
   @api
    b;
   
    get _value(){
        return this.a * this.b; // this will be called 4 times but the code is more compact
    } 
}
````

**Comment by gelliott181** [comment]
Sure, you can cache anything on the request side. It's not really something the framework needs to handle though, this could have been resolved with a StackOverflow question instead.

```
class Component extends LightningElement {
  cachedValue = expensiveCall();
  cacheTTL = 16; // Time unit is "Renders" because we tick the counter in the renderedCallback
  ticksSinceUpdate = 0;

  get value() {
    if (!this.cachedValue || this.ticksSinceLastUpdate > this.cacheTTL) {
      this.cachedValue = expensiveCall();
      this.ticksSinceLastUpdate = 0;
    }

    return this.cachedValue;
  }

  async tick() {
    this.ticksSinceUpdate++;
  }

  renderedCallback() {
    this.tick();
  }
```

**Comment by Bartheleway** [comment]
Hi, I think there is something to do with the rendering process to avoid such. On the js side, you can always temporarily cache a getter value if you want to loop over for example but we have no control when it comes to template side.

In your example I would move the `if:true` to the root div as it will make more sense. I understand that it's just to give an idea of the issue ;)

**Comment by AllanOricil** [comment]
@Bartheleway thank you for your suggestion. But that was just a simple example to show one point that the framework could be improved. I know I can solve it using the suggestions above, or even yours, but it would be good if lwc could offer something out of the box to keep the code simple, not repetitive and avoid unnecessary re-render automagically.

**Comment by gelliott181** [comment]
@AllanOricil I saw you didn't like my original solution, and I re-read the first post to see if I missed something. Turns out I did, and I have an even better answer: Wires.

Recently custom wire adapters were enabled on platform, and I've been working with them heavily. Here's a sample LWC that illustrates how we can take advantage of the wire service to avoid unnecessary recalculation of values when the component itself renders: https://webcomponents.dev/edit/cDH0vLumya6JnRqOqgfU

I've already introduced this pattern into an LWC codebase and it's brought significant performance improvements to components.
